File: sorting/clojure/src/toys/insertion_sort.clj
(ns
    ^{:doc "Implementation of insertion sort algorithm."
      :author "Paul Evans"}
    toys.insertion-sort)

(defn insert
  "Inserts value into sorted list.  Involves recursion yielding an iterative
process."
  [value list]
  (loop [list list
         newlist []]
    (if (empty? list)
      (vec (cons value newlist))
      (if (= (count list) 1)
        (if (< value (first list))
          (vec (cons value (vec (cons (first list) newlist))))
          (vec (conj (vec (cons (first list) newlist)) value)))
        (if (and (<= value (last list))
                 (> value (nth list (- (count list) 2))))
          (vec (concat (take (dec (count list)) list)
                       (cons value (cons (last list) newlist))))
          (if (> value (last list))
            (vec (conj list value))
            (recur (take (dec (count list)) list)
                   (cons (last list) newlist))))))))

(defn insertion-sort
  "Sorts list using insertion sort algorithm.  Involves recursion yielding
an interative process."
  [list]
  (if (or (empty? list)
          (= (count list) 1))
      list
      (loop [sorted-list (vector (first list))
             index 1]
        (if (= (count sorted-list) (count list))
          sorted-list
          (let [val (nth list index)]
            (recur (insert val sorted-list) (inc index)))))))

File: sorting/javascript/src/insertion-sort.js
/**
 * Inserts value into given sorted list, preserving sortedness.
 */
function insert(value, list) {
    var newList;
    var loop;
    var innerLoop;
    var valAdded;
    if (list == null || list.length == 0) {
        return [value];
    } else if (list.length == 1) {
        if (value >= list[0]) {
            return [list[0], value];
        } else {
            return [value, list[0]];
        }
    } else {
        valAdded = false;
        newList = new Array(list.length + 1);
        for (loop = 0; loop < list.length && !valAdded; loop++) {
            if (value < list[loop]) {
                newList[loop] = value;
                for (innerLoop = loop; innerLoop < list.length; innerLoop++) {                    
                    newList[innerLoop + 1] = list[innerLoop];
                }
                valAdded = true;
            } else {
                newList[loop] = list[loop];
            }
        }
        if (!valAdded) {
            newList[list.length] = value;
        }
        return newList;
    }
}

function insertionSort(list) {
    var sortedList;
    var value;
    var loopCounter;
    if (list == null || list.count === 0 || list.count === 1) {
        return list;
    } else {
        sortedList = [];
        sortedList[0] = list[0];
        loopCounter = 1;
        while (loopCounter < list.length) {
            value = list[loopCounter];
            sortedList = insert(value, sortedList);
            loopCounter++;
        }
        return sortedList;
    }
}